### ISSUE TITLE
How to measure distance of objects from the camera?

### ISSUE THREAD

**Issue by onkarganjewar** [rbgirshick/py-faster-rcnn#566]
Currently, I'm working on my [project](https://github.com/onkarganjewar/cmpe295-masters-project#object-and-lane-detection-for-self-driving-cars) that detects objects/lanes for self-driving-cars using py-faster-rcnn. So far, I'm able to locate and label(left/right) objects properly. I want to add the distance to the labels of the detected objects (cars) from the camera. Is there any way that I could be able to measure the distance of the detected objects using py-faster-rcnn? Any help/guidance is appreciated. Thanks.

**Comment by deboc** [comment]
Hi,
That's a really interesting project !
I think the most easy way to achieve it is to use the size of the bounding boxes from the deepnet and compute the distance afterward. You have to assume that the road is flat (no slope) and know a good calibration of the camera used for the images dataset. IMO that's how most of image base distance estimation is done in production.
But if you absolutely want to play with distance estimation with deep learning (and that's definitely more interesting !) I suggest you just provide the ground truth distance in the data layer and update the loss so the network tries to learn this feature in a end-to-end manner. In this case you need a dataset with the bounding box + distance in the annotation (like the [KITTI](http://www.cvlibs.net/datasets/kitti/) one)

**Comment by onkarganjewar** [comment]
@deboc Thanks for your response. Appreciate your insights. I think I would not go further with deep learning approach for the distance estimation as I fear it would make my algorithm even slower. The object detection part (py-faster-rcnn) is already consuming a significant amount of time in my project, so I'm just worried about the efficiency of my project if I were to run any more deepnets. 

Having said that, I'm currently trying to achieve a distance estimation in a similar manner as you have described with your first approach. However, I'm not able to exactly correlate the bounding box dimensions to the proposed distance. As of now, I'm attempting to get the top view of the road using the perspective transform. I'm not really sure how to go from here, or even get the accurate parameters for the perspective transform. I'm trying to fine tune the coordinates of the transformation matrix, but still, I'm not getting the satisfactorily results with the perspective transform for the videos of varying resolutions/sizes.  

If I have misconstrued any of the things that I've mentioned earlier, please do rectify them. Kindly pardon my lack of comprehension as I'm still a novice to this field. So, would you mind being more elaborate about your first approach(links to a couple of tutorials would greatly help), or provide any feedback for my current distance estimation approach, please? Thank you.

**Comment by harshilpatel312** [comment]
Late to the party but for future reference..

I used detection information (bounding boxes) and distance estimates to train a network. You can look it up [here](https://github.com/harshilpatel312/KITTI-distance-estimation)

**Comment by alokranjan007** [comment]
Hey @harshilpatel312, I looked on to your project, its really nice. I have couple of questions. 

Could you please help me to understand how you pass the information of location (x,y,z) in your annotation file? Are these location information of the object is known as priory?  For annotation of data in Iabelimage tool, there is no option where one can pass on the location information. 

Also, you have done this on Kitti dataset, do you have any suggestion for the custom data set? How one should train for the custom data set for distance estimation of the object.


@onkarganjewar Have you find any solution based  on the bounding box size for distance estimation?

**Comment by harshilpatel312** [comment]
The location information exists in the KITTI dataset. It's not known priori. KITTI used a suite of sensors to get that location information. 

For custom datasets, you need a way to measure these distances for ground truth somehow. For small robots, you can use gps or trackers. For bigger projects, if you have the resources, then LIDAR and Radar work best.

**Comment by NoMoreHiding27** [comment]
@harshilpatel312  
Is there any way to use Kitti dataset to re-train using transfer learning based on tensorflow object detection api?

**Comment by harshilpatel312** [comment]
I haven't tried it but I guess it should be pretty straightforward.

Get the images and annotations from KITTI dataset, pick your favorite model from Tensorflow Object Detection API and re-train it. Sentdex does an excellent job [here](https://www.youtube.com/watch?v=COlbP62-B-U&list=PLQVvvaa0QuDcNK5GeCQnxYnSSaar2tpku) demonstrating how to re-train the Tensorflow Object Detection models.
